processor/batchprocessor/batch_processor.go

```diff
@@ -0,0 +1,223 @@
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector/processor"
+)
+
+const (
+	initialBatchCapacity = uint32(16)
+)
+
+// batch_processor is a component that accepts spans, places them into batches and sends downstream.
+//
+// batch_processor implements consumer.TraceConsumer
+//
+// Batches are sent out with any of the following conditions:
+// - batch size reaches cfg.SendBatchSize
+// - cfg.Timeout is elapsed since the timestamp when the previous batch was sent out.
+type batchProcessor struct {
+	sender consumer.TraceConsumer
+	name   string
+	logger *zap.Logger
+
+	batch  *batch
+	ticker *ticker
+
+	sendBatchSize uint32
+	tickTime      time.Duration
+	timeout       time.Duration
+}
+
+var _ consumer.TraceConsumer = (*batchProcessor)(nil)
+
+// newBatchProcessor creates a new batcher that batches spans by node and resource
+func newBatchProcessor(params component.ProcessorCreateParams, sender consumer.TraceConsumer, cfg *Config) *batchProcessor {
+	bp := &batchProcessor{
+		name:   cfg.Name(),
+		sender: sender,
+		logger: params.Logger,
+
+		sendBatchSize: cfg.SendBatchSize,
+		tickTime:      cfg.TickTime,
+		timeout:       cfg.Timeout,
+	}
+
+	bp.batch = newBatch(bp)
+
+	// start ticker after options loaded in
+	bp.ticker = newTicker(bp)
+	go bp.ticker.start()
+
+	return bp
+}
+
+// ConsumeTrace implements batcher as a SpanProcessor and takes the provided spans and adds them to
+// batches
+func (bp *batchProcessor) ConsumeTrace(ctx context.Context, td data.TraceData) error {
+	bp.batch.add(td)
+	return nil
+}
+
+func (bp *batchProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: true}
+}
+
+// Start is invoked during service startup.
+func (bp *batchProcessor) Start(ctx context.Context, host component.Host) error {
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (bp *batchProcessor) Shutdown(context.Context) error {
+	// TODO: flush accumulated data.
+	return nil
+}
+
+func (bp *batchProcessor) sendItems(
+	itemsToProcess []data.TraceData,
+	resourceSpansCount uint32,
+	measure *stats.Int64Measure,
+) {
+	td := data.NewTraceData()
+	rss := td.ResourceSpans()
+
+	// TODO: Use resourceSpansCount to set capacity of the rss slice
+	// For now there is no way to do that with internal data structure API
+
+	for _, td := range itemsToProcess {
+		td.ResourceSpans().MoveTo(rss)
+	}
+
+	statsTags := []tag.Mutator{tag.Insert(processor.TagProcessorNameKey, bp.name)}
+	_ = stats.RecordWithTags(context.Background(), statsTags, measure.M(1))
+
+	// TODO: This process should be done in an async way, perhaps with a channel + goroutine worker(s)
```

Comment:
Why does it need in to be async?

Reply:
This TODO is carried over. do you think should not consider running the next consumer call in goroutine?